src/internal/arg.c

```diff
@@ -84,3 +84,98 @@ sexp* rlang_is_missing(sexp* _call, sexp* _op, sexp* args, sexp* env) {
 
   return r_lgl(r_eval(r_x_sym, env) == r_missing_sym);
 }
+
+static sexp* stop_arg_match_call = NULL;
+void arg_match0_abort(const char* msg, sexp* arg_nm_promise, sexp* env);
+sexp* rlang_ext2_arg_match0(sexp* _call, sexp* _op, sexp* args, sexp* env) {
+  args = r_node_cdr(args);
+
+  sexp* arg =            KEEP(r_eval(r_node_car(args), env)); args = r_node_cdr(args);
+  sexp* values =         KEEP(r_eval(r_node_car(args), env)); args = r_node_cdr(args);
+  // Expensive by default, forced only if necessary.
+  sexp* arg_nm_promise = r_node_car(args);
+
+  if (r_typeof(arg) != r_type_character) {
+    arg_match0_abort("`%s` must be a character vector.", arg_nm_promise, env);
+  }
+  if (r_typeof(values) != r_type_character) {
+    r_abort("`values` must be a character vector.");
+  }
+
+  r_ssize len = r_length(arg);
+  r_ssize len_values = r_length(values);
+  if (len != 1 && len != len_values) {
+    //arg_match0_abort("`%s` must be a string or have the same length as `values`.",  arg_nm_promise, env);
+
+    arg = Rf_ScalarString(STRING_ELT(arg, 0));
```

Comment:
From memory there should be something like `r_str_as_character(r_chr_get(arg, 0))`

Reply:
Same for the other usages of STRING_ELT and ScalarString below.